ppx/query.mll

```diff
@@ -0,0 +1,113 @@
+{
+open Ppx_mysql_runtime.Stdlib
+
+type param =
+  { typ : string
+  ; opt : bool
+  ; name : string
+  ; of_string : string * string
+  ; to_string : string * string }
+
+type parsed_query =
+  { query : string
+  ; in_params : param list
+  ; out_params : param list }
+
+type parse_error =
+  [ `Bad_identifier of string
+  | `Unknown_type_spec of string
+  | `Unterminated_string
+  | `Escape_at_end ]
+
+let build_param spec opt name =
+  let open Result in
+  begin match spec with
+  | "int" -> Ok ("int", ("Pervasives", "int_of_string"), ("Pervasives", "string_of_int"))
+  | "int32" -> Ok ("int32", ("Int32", "of_string"), ("Int32", "to_string"))
+  | "int64" -> Ok ("int64", ("Int64", "of_string"), ("Int64", "to_string"))
+  | "string" -> Ok ("string", ("Ppx_mysql_runtime", "identity"), ("Ppx_mysql_runtime", "identity"))
+  | spec -> Error (`Unknown_type_spec spec)
+  end >>= fun (typ, of_string, to_string) ->
+  Ok {typ; opt = (opt = "?"); name; of_string; to_string}
+}
+
+let escape = '\\'
+let squot = '\''
+let dquot = '"'
+let quot = squot | dquot
+let digit = ['0'-'9']
+let lower = ['a'-'z']
+let upper = ['A'-'Z']
+let underscore = '_'
+let ident = (lower | underscore) (lower | upper | underscore | digit)*
+let spec = (lower | upper | underscore | digit)+
+
+rule main_parser buf acc_in acc_out = parse
+  | quot as delim
+    {Buffer.add_char buf delim;
+    quotation_parser buf acc_in acc_out delim lexbuf}
+  | '%' (spec as spec) ('?'? as opt) '{' ([^ '}' ]+ as name) '}'
+    {let open Result in
+    check_ident name (Lexing.from_string name) >>= fun () ->
+    build_param spec opt name >>= fun in_param ->
+    Buffer.add_char buf '?';
+    main_parser buf (in_param :: acc_in) acc_out lexbuf}
+  | '@' (spec as spec) ('?'? as opt) '{' ([^ '}' ]+ as name) '}'
```

Comment:
Same here, I think `name` could be constrained to valid OCaml identifiers.

Reply:
Note that in the case of output parameters we do not want any constraints at all, because the underlying `name` may not be just a column name but in fact an arbitrary SQL expression. As an example, the following is (and ought to remain) perfectly legal: `SELECT @int{age+1} FROM users`.